### PR TITLE
added OOP approach for BST

### PR DESCRIPTION
**Pull request by Dharni0607**
Added object oriented approach for BST. Please review and let me know if any changes are required.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/966"></a>

### CODE REVIEW

**Comment by kvedala** [comment]
Thank you for your contribution 👍 
Please note the following:
1. Please do not delete the author checks in the description and add appropriate description. It is meant to facilitate contribution by authors to maintain repo standards
2. please ensure the self-tests pass
3. please ensure to read the contribution guidelines - code needs to be documented per doxygen standards to ensure that a layman can read and learn from your code
4. Please ensure to fix the alerts reported by LGTM.

**Comment by Dharni0607** [comment]
Ok. I will make the changes.

**Comment by kvedala** [comment]
> Ok. I will make the changes.

Thank you. Please review code in [this file](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/data_structures/skip_list.cpp) as well for reference.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 1]
```diff
@@ -0,0 +1,205 @@
+/**
+ * \file
+ * \brief A simple tree implementation using structured nodes
+ * \Object Oriented approach
+ */
+
+#include<bits/stdc++.h> 
+#include<queue>
+using namespace std;
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:9]: Please remove this and use `std::` where necessary.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
+/**
+ * \file
+ * \brief A simple tree implementation using structured nodes
+ * \Object Oriented approach
+ */
+
+#include<bits/stdc++.h>
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:7]: Don't use this library as it slows down the compilation process and is quite Linux-specified.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+ * \file
+ * \brief A simple tree implementation using structured nodes
+ * \Object Oriented approach
+ */
+
+#include<bits/stdc++.h> 
+#include<queue>
+using namespace std;
+
+struct node {
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:11]: Undocumented struct.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+class BinarySearchTree{
+    node *root;
+
+public:
+    BinarySearchTree(){
+        root = NULL;
+    }
+
+    void insert(int data){
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:25]: Undocumented class function.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        node *newnode = new node;
+        newnode->value = data;
+        newnode->left = NULL;
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:28]: Please use `nullptr` here, and in other parts of the code.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+                    current = current->left;
+            }
+            else{
+                if(current->right == NULL)
+                {
+                    current->right = newnode;
+                    return;
+                }
+                else
+                    current = current->right;
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:52]: Make it look like C++ code:
```suggestion
                else {
                    current = current->right;
                }
```

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+            }
+            
+        } 
+    }
+
+    node *getRoot(){
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:58]: Undocumented.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+        return root;
+    }
+    
+    void BFT() {
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:62]: Undocumented function.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+            q.pop();
+            cout << current->value << " ";
+            if(current->left)
+                q.push(current->left);
+            if(current->right)
+                q.push(current->right);
+        }
+    }
+
+    void preOrder(node *current){
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:76]: Same here.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+    node *minValueNodeInRightST(node *node) 
+    { 
+        struct node *current = node; 
+        while (current && current->left) 
+            current = current->left;       
+        return current; 
+    } 
+    
+    node *deleteNode(node *current, int target){
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:111]: Same here and other functions above.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+        }
+        return current;
+    }
+    
+    void remove(int target){
+        root = deleteNode(root, target);
+    }
+};
+
+int main() {
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:157]: Please add a comment structure like:

```c++
/**
 * Main function
 */
```

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+            break;
+        case 4:
+            BST.preOrder(BST.getRoot());
+            break;
+        case 5:
+            BST.inOrder(BST.getRoot());
+            break;
+        case 6:
+            BST.postOrder(BST.getRoot());
+            break;
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:200]: Missing `default` statement.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+class BinarySearchTree{
+    node *root;
+
+public:
+    BinarySearchTree(){
+        root = NULL;
+    }
+
+    void insert(int data){
+        node *newnode = new node;
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:26]: Allocated memory is not freed/released.

File: data_structures/binary_search_tree_OOP_approach.cpp [hunk 14]
```diff
@@ -0,0 +1,205 @@
+/**
+ * \file
+ * \brief A simple tree implementation using structured nodes
+ * \Object Oriented approach
+ */
```
Comment (Panquesito7) [inline data_structures/binary_search_tree_OOP_approach.cpp:5]: As @kvedala mentioned, please update this comment block like in [`skip_list`](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/data_structures/skip_list.cpp#L1-#L14).

**Comment by Panquesito7** [comment]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/893869500) warnings.

**Comment by kvedala** [comment]
> Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/893869500) warnings.

Please, avoid posting multiple reviews until at least one of them is complete. Multiple reviews confuse the authors and create an unnecessary branching of fixes.

File: data_structures/binary_search_tree_OOP_approach.cpp (hunk 13 above)
Comment (kvedala) [reply data_structures/binary_search_tree_OOP_approach.cpp:58]: it is being release appropriately.

**Comment by Dharni0607** [comment]
Hi @kvedala @Panquesito7 , I have updated the PR, there are couple of errors and warnings for destructor method and deletion of node. Can you please guide me.

**Comment by kvedala** [comment]
errors:
1. `
/home/runner/work/C-Plus-Plus/C-Plus-Plus/data_structures/binary_search_tree_oop_approach.cpp:25:7: error: class 'BinarySearchTree' defines a non-default destructor but does not define a copy constructor, a copy assignment operator, a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions,-warnings-as-errors]`
This error asks to create a copy or assignment operator to the class. You can define one as 
```cpp
class BinarySearchTree {
public:
    BinarySearchTree(BinarySearchTree const&) = delete;
    BinarySearchTree& operator =(BinarySearchTree const&) = delete;
}
```
as class member functions
2. this can be avoided by the second suggestion made: Use `std::shared_ptr` or `std::unique_ptr` functions of *smart pointers* feature of C++11 that will take care of memory de-allocation and thus make the destructor function useless.

Without the use of smart pointers, the implementation is so rudimentary that it can be implemented almost verbatim as a C program.

**Comment by kvedala** [comment]
The error mentions that either
* both copy constructor and the destructor should be defined 
* or neither should be custom defined

**Comment by kvedala** [comment]
1. Missing move constructor:
```cpp
BinarySearchTree(BinarySearchTree const&&) = delete;
```

2. Error on line 81: `/home/runner/work/C-Plus-Plus/C-Plus-Plus/data_structures/binary_search_tree_oop_approach.cpp:157:41: error: parameter 'n' is passed by value and only copied once; consider moving it to avoid unnecessary copies [performance-unnecessary-value-param,-warnings-as-errors]`

**Comment by Panquesito7** [comment]
@Dharni0607 open the Gitpod link above and open the file you added.
There are some `clang-tidy` suggestions and quick-fixes there.

If you are still a bit confused, I can make a review fixing various `clang-tidy` warnings. 🙂

**Comment by Panquesito7** [comment]
No response from author.
Closing this pull request.

If you are going to re-work on this, please ping me and I'll re-open the PR.

**Pull request closed without merging**